collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -153,40 +159,49 @@ func (h *headerReceiver) combineHeaders(ctx context.Context, hdrsBytes []byte) (
 	}
 
 	// Note that we will parse the headers even if they are not
-	// used, to check for validity.  tmpHdrsAppend() will skip
-	// modifying tmpHdrs if it is nil.
-	h.tmpHdrs = nil
-
-	needMergedHeaders := h.includeMetadata || h.hasAuthServer
-
-	// If headers are being merged, allocate a new map.
-	if needMergedHeaders {
-		h.tmpHdrs = map[string][]string{}
-	}
+	// used, to check for validity and/or trace context.
+	h.tmpHdrs = map[string][]string{}
 
 	// Write calls the emitFunc, appending directly into `tmpHdrs`.
 	if _, err := h.decoder.Write(hdrsBytes); err != nil {
 		return ctx, nil, err
 	}
 
-	if needMergedHeaders {
-		// Add streamHdrs that were not carried in the per-request headers.
-		for k, v := range h.streamHdrs {
-			// Note: This is done after the per-request metadata is defined
-			// in recognition of a potential for duplicated values stemming
-			// from the Arrow exporter's independent call to the Auth
-			// extension's GetRequestMetadata().  This paired with the
-			// headersetter's return of empty-string values means, we would
-			// end up with an empty-string element for any headersetter
-			// `from_context` rules b/c the stream uses background context.
-			// This allows static headers through.
-			//
-			// See https://github.com/open-telemetry/opentelemetry-collector/issues/6965
-			lk := strings.ToLower(k)
-			if _, ok := h.tmpHdrs[lk]; !ok {
-				h.tmpHdrs[lk] = v
+	// Extract trace context, if present, from the per-request
+	// headers before stream context values are added.  First
+	// check if tracing is enabled, since allocations will be
+	// required to turn the multi-string-valued headers map into a
+	// single-string-value map.
+	carrier := otel.GetTextMapPropagator()
+	if len(carrier.Fields()) != 0 {
+		flat := map[string]string{}
+		for _, key := range carrier.Fields() {
+			have := h.tmpHdrs[key]
+			if len(have) > 0 {
+				flat[key] = have[0]
+				delete(h.tmpHdrs, key)
 			}
 		}
+
+		ctx = carrier.Extract(ctx, propagation.MapCarrier(flat))
+	}
+
+	// Add streamHdrs that were not carried in the per-request headers.
+	for k, v := range h.streamHdrs {
```

Comment:
Hmm this loop used to be guarded by `needMergedHeaders := h.includeMetadata || h.hasAuthServer`, any reason we don't need to check this condition anymore?

Reply:
I think the code as it stood was a little over-complicated as it stood, and it would become more complicated if you added another clause "or the context is traced" to the list of reasons why we might need to allocate in this function.

I recognize that I'm splitting hairs here--in this case I removed some (complex) logic and accept the extra allocations. In the other comment of yours, I've preserved a (simple) conditional that avoids allocations.